sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobAsyncClientBase.java

```diff
@@ -767,29 +745,201 @@ private AsynchronousFileChannel downloadToFileResourceSupplier(String filePath)
+        int numBlocks = toIntExact(dataSize / blockLength);
+        // Include an extra block for trailing data.
+        if (dataSize % blockLength != 0) {
+            numBlocks++;
         }
-        long offset = blobRange.getOffset();
-        long length = blobRange.getCount();
-        List<BlobRange> chunks = new ArrayList<>();
-        for (long pos = offset; pos < offset + length; pos += blockSize) {
-            long count = blockSize;
-            if (pos + count > offset + length) {
-                count = offset + length - pos;
+        return numBlocks;
+    }
+
+    /*
+    Download the first chunk. Construct a Mono which will emit the total count for calculating the number of chunks,
+    access conditions containing the etag to lock on, and the response from downloading the first chunk.
+     */
+    private Mono<Tuple3<Long, BlobRequestConditions, BlobDownloadAsyncResponse>> getSetupMono(BlobRange r,
+        ParallelTransferOptions o, DownloadRetryOptions downloadRetryOptions,
+        BlobRequestConditions requestConditions, boolean rangeGetContentMd5, Context context) {
+        // We will scope our initial download to either be one chunk or the total size.
+        long initialChunkSize = r.getCount() != null && r.getCount() < o.getBlockSize()
+            ? r.getCount() : o.getBlockSize();
+
+        return this.downloadWithResponse(new BlobRange(r.getOffset(), initialChunkSize), downloadRetryOptions,
+            requestConditions, rangeGetContentMd5, context)
+            .subscribeOn(Schedulers.elastic())
+            .flatMap(response -> {
+                /*
+                Either the etag was set and it matches because the download succeeded, so this is a no-op, or there
+                was no etag, so we set it here. ETag locking is vital to ensure we download one, consistent view
+                of the file.
+                 */
+                BlobRequestConditions newConditions = setEtag(requestConditions, response.getHeaders()
+                    .getValue("ETag"));
+
+                // Extract the total length of the blob from the contentRange header. e.g. "bytes 1-6/7"
+                long totalLength = extractTotalBlobLength(response.getHeaders().getValue("Content-Range"));
+
+                /*
+                If the user either didn't specify a count or they specified a count greater than the size of the
+                remaining data, take the size of the remaining data. This is to prevent the case where the count
+                is much much larger than the size of the blob and we could try to download at an invalid offset.
+                 */
+                long newCount = r.getCount() == null || r.getCount() > (totalLength - r.getOffset())
+                    ? totalLength - r.getOffset() : r.getCount();
+
+                return Mono.zip(Mono.just(newCount), Mono.just(newConditions), Mono.just(response));
+            })
+            .onErrorResume(throwable -> {
```

Comment:
Give that we should only be resuming if the exception is a `BlobStorageException` could we use the `onErrorResume` that takes a `Class<E>` to simplify the check logic so we don't need to verify instance and perform casting?

https://projectreactor.io/docs/core/release/api/reactor/core/publisher/Mono.html#onErrorResume-java.lang.Class-java.util.function.Function-

Reply:
I didn't use that overload because I need additional logic on top of just validating the type anyway, but if you think it'll be cleaner, I can change it

Reply:
Under the hood the method is the same as the predicate version where the predicate is `Class.isInstance(exception)` then the exception passed in is the type of the class.